### ISSUE TITLE
Add posibility to override context classes

### ISSUE THREAD

**Issue by torinaki** [schmittjoh/JMSSerializerBundle#746]
I have faced with a need to override default Deserialization/Serialization context classes. In my particular case, I would like to use store some meta information produced by custom Handlers.

Currently, it is not possible because of methods `createDeserializationContext` and `createSerializationContext` of class `\Mango\Bundle\JsonApiBundle\ContextFactory\ConfiguredContextFactory` rely on private method `configureContext` which contains configuration logic (I just tried to override factory class)

Is it possible to provide such a feature?

**Comment by goetas** [comment]
Can't the setAttribute and getAttribute do the job, instead of creating new classes?

**Comment by torinaki** [comment]
@goetas I'm not aware of all possible keys in attributes

**Comment by torinaki** [comment]
@goetas when a handler is getting the context it already is initialized, and attributes are in a read-only state

**Comment by torinaki** [comment]
@goetas my problem might be solved by introducing some metadata storage in context. It is exactly what I'm going to add into Context class.
IMO context attributes should stay immutable.

**Comment by goetas** [comment]
So your idea is to have handlers setting some data in the context, and retrieve them later, right?

**Comment by torinaki** [comment]
@goetas, it seems I have found a way. During serialization each handler getting the visitor object. In my case visitor already is overridden. So my initial idea isn’t actual anymore.

To give you more context I’m working on implementing JSON API specification. Unfortunately, the JSON API document structure doesn't look like a tree it is flat. I need to enrich output array on root level sometimes and this contradicts with JMS library approaches which assume that we want to get a tree on the output.
Here is a PR with JMS 2.0 support: https://github.com/ecentria/JsonApiBundle/pull/15

I will close this ticket due to I have found the way to implement it.
